src/main/java/com/codejoust/main/service/GameManagementService.java

```diff
@@ -265,7 +266,24 @@ public void updateCode(String roomId, String userId, PlayerCode playerCode) {
         liveGameService.updateCode(game.getPlayers().get(userId), playerCode);
     }
 
-    protected void endGame(Game game) {
+    public GameDto manuallyEndGame(String roomId, EndGameRequest request) {
+        Game game = getGameFromRoomId(roomId);
+
+        User initiator = UserMapper.toEntity(request.getInitiator());
+        if (!game.getRoom().getHost().equals(initiator)) {
+            throw new ApiException(GameError.INVALID_PERMISSIONS);
+        }
+
+        game.setGameEnded(true);
+        handleEndGame(game);
+
+        GameDto gameDto = GameMapper.toDto(game);
+        socketService.sendSocketUpdate(gameDto);
+
+        return gameDto;
+    }
```

Comment:
Does there have to be any handling on if the game is already ended and this method is called?

Also, I will note that I thought there could arise an issue where the game ends by timer and by manual end game at the same time, but when I tried testing this out by ending the game at around one second, no adverse consequences seemed to occur. 👍

Reply:
I don't believe that would cause any issues, since it's just setting that variable to true and canceling the timers - rn I think the one verification we do is if trying to play again when the game hasn't ended yet

However this brings up a good point cause (unless I'm completely missing it) I don't think we do any validation on starting a game or submitting solutions, so that could probably totally screw things up lol. Probably a lower priority since it's an edge case but something to consider

Reply:
> I don't believe that would cause any issues, since it's just setting that variable to true and canceling the timers - rn I think the one verification we do is if trying to play again when the game hasn't ended yet

That makes sense - if we ever have statistics or show on the UI how the game ended (out of time, ended game early, everybody submitted), then we may need to update this method - otherwise this probably isn't an issue.

Reply:
I think like preventing users from submitting after the game is over, or from starting a game when it's already active, etc.